File: backend-elixir/lib/backend_elixir/external/query.ex
defmodule BackendElixir.External.Query do
  @moduledoc """
  外部公開API(GET /external/v1/tasks)のクエリパラメータ解析(ページング方式のパラメータ変換・
  境界値の丸め込み)を、HTTPハンドラから独立した単体テスト可能な関数として切り出す
  """

  alias BackendElixir.Domain.TaskError

  @doc """
  user_idクエリパラメータを解析する。空/欠如は:user_id_required、数値でなければ:invalid_user_id
  """
  def parse_user_id(params) do
    case Map.get(params, "user_id") do
      nil -> {:error, TaskError.user_id_required()}
      "" -> {:error, TaskError.user_id_required()}
      raw ->
        case Integer.parse(raw) do
          {user_id, ""} -> {:ok, user_id}
          _ -> {:error, TaskError.invalid_user_id()}
        end
    end
  end

  @doc "v1(offset)向け。page/page_sizeいずれも1未満は1に丸める"
  def parse_offset_page(params) do
    page = params |> Map.get("page") |> parse_positive_int(1)
    page_size = params |> Map.get("page_size") |> parse_positive_int(10)
    {page, page_size}
  end

  @doc "v2(cursor)向け。cursorは数値でなければ先頭からとみなす(after_id=0)。limitは1未満は1に丸める"
  def parse_cursor_page(params) do
    after_id =
      case Map.get(params, "cursor") do
        nil -> 0
        "" -> 0
        raw ->
          case Integer.parse(raw) do
            {value, ""} -> value
            _ -> 0
          end
      end

    limit = params |> Map.get("limit") |> parse_positive_int(10)
    {after_id, limit}
  end

  defp parse_positive_int(nil, default), do: default
  defp parse_positive_int("", default), do: default

  defp parse_positive_int(raw, default) do
    case Integer.parse(raw) do
      {value, ""} when value >= 1 -> value
      _ -> default
    end
  end
end

File: backend-haskell/src/BackendHaskell/External/Query.hs
-- | 外部公開API(/external/v1/tasks)のクエリパラメータ解析。DB・HTTPを一切介さない
-- 純粋関数として分離しており、単体テストで直接検証できる(backend-c/backend-cppの
-- external_query相当)
module BackendHaskell.External.Query
  ( parseUserId
  , offsetPage
  , offsetPageSize
  , cursorAfterId
  , cursorLimit
  ) where

import Data.Text (Text)
import qualified Data.Text as T
import Text.Read (readMaybe)

import BackendHaskell.Domain.Errors (TaskError (..), TaskErrorKind (InvalidRequest))

-- | user_idは必須。欠落/空文字はuser_id_required、数値でなければinvalid_user_id
parseUserId :: Maybe Text -> Either TaskError Int
parseUserId Nothing = Left (TaskError InvalidRequest "user_id_required")
parseUserId (Just raw)
  | T.null raw = Left (TaskError InvalidRequest "user_id_required")
  | otherwise = case readMaybe (T.unpack raw) :: Maybe Int of
      Just n -> Right n
      Nothing -> Left (TaskError InvalidRequest "invalid_user_id")

-- | pageは1未満なら1にクランプする(既定1)
offsetPage :: Maybe Int -> Int
offsetPage = clampAtLeast1 . maybe 1 id

-- | page_sizeは1未満なら1にクランプする(既定10)
offsetPageSize :: Maybe Int -> Int
offsetPageSize = clampAtLeast1 . maybe 10 id

-- | cursor省略時は先頭から(after_id=0扱い)。0以下は「未指定」として扱う
cursorAfterId :: Maybe Text -> Int
cursorAfterId Nothing = 0
cursorAfterId (Just raw) = maybe 0 (\n -> if n > 0 then n else 0) (readMaybe (T.unpack raw) :: Maybe Int)

-- | limitは1未満なら1にクランプする(既定10)
cursorLimit :: Maybe Int -> Int
cursorLimit = clampAtLeast1 . maybe 10 id

clampAtLeast1 :: Int -> Int
clampAtLeast1 n = max 1 n
